File: leetcode/318_maximum_product_of_word_lengths.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) 
    {
        vector<bitset<26>> bits(words.size());
        for(int i=0;i<words.size();++i)
        {
            bits[i].reset();
            for(int j=0;j<words[i].length();++j)
            {
                bits[i].set(words[i][j]-'a');
            }
        }
        
        bitset<26> t(0);   
        int max_ans =0; 
        for(int i=0;i<words.size();++i)
        {
            for(int j=i+1;j<words.size();++j)
            {
                bitset<26> curr;
                curr = bits[i];
                curr &= bits[j];
                if(curr == t)
                {
                    if(max_ans < words[i].length()*words[j].length())
                    {
                        max_ans = words[i].length()*words[j].length();
                    }
                }
            }
        }
        return max_ans;
    }
};

File: leetcode/318_maximum_product_of_word_lengths.java
class Solution {
    public int maxProduct(String[] words) 
    {
       
        
        int[] bitset = new int[words.length];
        for(int i=0;i<words.length;++i )
        {
            String word = words[i];
            for(int j=0;j<word.length();++j)
            {
                bitset[i] |= 1<<(word.charAt(j)-'a');
            }
        }
        int ans = Integer.MIN_VALUE;
        for(int i=0;i<words.length;++i)
        {
            for(int j=i+1;j<words.length;++j)
            {
                if((bitset[i]&bitset[j])==0)
                {
                    ans = Math.max(ans,words[i].length()* words[j].length());
                }
            }
        }
        return (ans==Integer.MIN_VALUE)?0:ans;
    }
}
